### PR TITLE
upgrade test special case comparing tag versions

### PR DESCRIPTION
**Pull request by knifewine**
When x.y.z has been tagged it should always win over x.y.z-foo

### CODE REVIEW

**Comment by knifewine** [comment]
er wait, don't merge, need to fix it up a little.

**Comment by knifewine** [comment]
force pushed a little fix, ready for review now.

File: upgrade_through_versions_test.py [hunk 1]
```diff
@@ -92,6 +92,12 @@ def __init__(self, git_ref, semver_str):
             self.semver = LooseVersion(make_ver_str(TRUNK_VER))
 
     def __cmp__(self, other):
+        # if we find a tag of x.y.z we need to special case that to compare higher than x.y.z-foo
+        # to accomplish this, compare on length and then invert so shorter version wins
+        if (len(self.semver.version) == 3) or (len(other.semver.version) == 3):
```
Comment (mambocab) [inline upgrade_through_versions_test.py]: Will this be thrown off by the 3.X series for X >= 1, which has base version numbers with no third component? What if, e.g., there's a 3.2-beta release? I don't know if these are planned, hopefully we'll just cut releases monthly, but who knows.

**Comment by mambocab** [comment]
Done with review. Code looks good, but I can't tell if the version numbers for 3.X where X>0 have been considered.

**Comment by knifewine** [comment]
I hadn't considered x.y type versions. Had to rework to make that sensible. It's still kinda gross though.

In a comment below I'll paste some output showing how the comparisons would function on the current tags.

**Comment by knifewine** [comment]
indented output shows how comparisons would work with this latest commit:

comparing 2.1.0-rc7 and 2.1.0-deb
comparing 2.1.0-rc6 and 2.1.0-rc7
comparing 2.1.0-rc6 and 2.1.0-deb
comparing 2.1.0-rc1 and 2.1.0-rc6
comparing 2.1.0-rc1 and 2.1.0-deb
comparing 2.1.0-rc3 and 2.1.0-rc6
comparing 2.1.0-rc3 and 2.1.0-rc1
comparing 2.1.11 and 2.1.0-rc3
comparing 2.1.11 and 2.1.0-rc7
comparing 2.1.10 and 2.1.0-rc6
comparing 2.1.10 and 2.1.11
comparing 2.1.10 and 2.1.0-rc7
comparing 2.1.9 and 2.1.0-rc6
comparing 2.1.9 and 2.1.10
comparing 2.1.9 and 2.1.0-rc7
comparing 2.1.8 and 2.1.0-rc7
comparing 2.1.8 and 2.1.10
comparing 2.1.8 and 2.1.9
comparing 2.1.5 and 2.1.0-rc7
comparing 2.1.5 and 2.1.10
comparing 2.1.5 and 2.1.9
comparing 2.1.5 and 2.1.8
comparing 2.1.4 and 2.1.5
comparing 2.1.4 and 2.1.0-rc3
comparing 2.1.4 and 2.1.0-rc7
comparing 2.1.7 and 2.1.4
comparing 2.1.7 and 2.1.9
comparing 2.1.7 and 2.1.8
comparing 2.1.7 and 2.1.5
comparing 2.1.6 and 2.1.5
comparing 2.1.6 and 2.1.9
comparing 2.1.6 and 2.1.8
comparing 2.1.6 and 2.1.7
comparing 2.1.1 and 2.1.5
comparing 2.1.1 and 2.1.0-rc6
comparing 2.1.1 and 2.1.4
comparing 2.1.1 and 2.1.0-rc7
comparing 2.1.0 and 2.1.5
comparing 2.1.0 and 2.1.0-rc6
   special compare result: 1
comparing 2.1.0 and 2.1.1
comparing 2.1.0 and 2.1.0-rc7
   special compare result: 1
comparing 2.1.3 and 2.1.4
comparing 2.1.3 and 2.1.0-rc6
comparing 2.1.3 and 2.1.0
comparing 2.1.3 and 2.1.1
comparing 2.1.2 and 2.1.4
comparing 2.1.2 and 2.1.0-rc7
comparing 2.1.2 and 2.1.1
comparing 2.1.2 and 2.1.3
comparing 2.1.0-rc5 and 2.1.3
comparing 2.1.0-rc5 and 2.1.0-rc7
comparing 2.1.0-rc5 and 2.1.0-rc3
comparing 2.1.0-rc5 and 2.1.0-rc6
comparing 2.1.0-rc4 and 2.1.3
comparing 2.1.0-rc4 and 2.1.0-rc6
comparing 2.1.0-rc4 and 2.1.0-rc3
comparing 2.1.0-rc4 and 2.1.0-rc5
comparing 2.1.0-rc2 and 2.1.2
comparing 2.1.0-rc2 and 2.1.0-rc5
comparing 2.1.0-rc2 and 2.1.0-rc3
comparing 2.1.0-rc2 and 2.1.0-rc1
comparing 2.1.0-beta2 and 2.1.2
comparing 2.1.0-beta2 and 2.1.0-rc5
comparing 2.1.0-beta2 and 2.1.0-rc2
comparing 2.1.0-beta2 and 2.1.0-rc1
comparing 2.1.0-beta2 and 2.1.0-deb
comparing 2.1.0-beta1 and 2.1.1
comparing 2.1.0-beta1 and 2.1.0-rc4
comparing 2.1.0-beta1 and 2.1.0-rc1
comparing 2.1.0-beta1 and 2.1.0-deb
comparing 2.1.0-beta1 and 2.1.0-beta2
comparing 3.0.0-beta1 and 3.0.0-beta2
comparing 3.0.0-rc1 and 3.0.0-beta2
comparing 3.0.0-alpha1 and 3.0.0-beta2
comparing 3.0.0-alpha1 and 3.0.0-beta1
comparing 3.0.0-rc2 and 3.0.0-beta2
comparing 3.0.0-rc2 and 3.0.0-rc1
comparing 3.0.0 and 3.0.0-beta2
   special compare result: 1
comparing 3.0.0 and 3.0.0-rc2
   special compare result: 1
comparing 2.1.0-rc7 and 2.1.0-deb
comparing 2.1.0-rc6 and 2.1.0-rc7
comparing 2.1.0-rc6 and 2.1.0-deb
comparing 2.1.0-rc1 and 2.1.0-rc6
comparing 2.1.0-rc1 and 2.1.0-deb
comparing 2.1.0-rc3 and 2.1.0-rc6
comparing 2.1.0-rc3 and 2.1.0-rc1
comparing 2.1.11 and 2.1.0-rc3
comparing 2.1.11 and 2.1.0-rc7
comparing 2.1.10 and 2.1.0-rc6
comparing 2.1.10 and 2.1.11
comparing 2.1.10 and 2.1.0-rc7
comparing 2.1.9 and 2.1.0-rc6
comparing 2.1.9 and 2.1.10
comparing 2.1.9 and 2.1.0-rc7
comparing 2.1.8 and 2.1.0-rc7
comparing 2.1.8 and 2.1.10
comparing 2.1.8 and 2.1.9
comparing 2.1.5 and 2.1.0-rc7
comparing 2.1.5 and 2.1.10
comparing 2.1.5 and 2.1.9
comparing 2.1.5 and 2.1.8
comparing 2.1.4 and 2.1.5
comparing 2.1.4 and 2.1.0-rc3
comparing 2.1.4 and 2.1.0-rc7
comparing 2.1.7 and 2.1.4
comparing 2.1.7 and 2.1.9
comparing 2.1.7 and 2.1.8
comparing 2.1.7 and 2.1.5
comparing 2.1.6 and 2.1.5
comparing 2.1.6 and 2.1.9
comparing 2.1.6 and 2.1.8
comparing 2.1.6 and 2.1.7
comparing 2.1.1 and 2.1.5
comparing 2.1.1 and 2.1.0-rc6
comparing 2.1.1 and 2.1.4
comparing 2.1.1 and 2.1.0-rc7
comparing 2.1.0 and 2.1.5
comparing 2.1.0 and 2.1.0-rc6
   special compare result: 1
comparing 2.1.0 and 2.1.1
comparing 2.1.0 and 2.1.0-rc7
   special compare result: 1
comparing 2.1.3 and 2.1.4
comparing 2.1.3 and 2.1.0-rc6
comparing 2.1.3 and 2.1.0
comparing 2.1.3 and 2.1.1
comparing 2.1.2 and 2.1.4
comparing 2.1.2 and 2.1.0-rc7
comparing 2.1.2 and 2.1.1
comparing 2.1.2 and 2.1.3
comparing 2.1.0-rc5 and 2.1.3
comparing 2.1.0-rc5 and 2.1.0-rc7
comparing 2.1.0-rc5 and 2.1.0-rc3
comparing 2.1.0-rc5 and 2.1.0-rc6
comparing 2.1.0-rc4 and 2.1.3
comparing 2.1.0-rc4 and 2.1.0-rc6
comparing 2.1.0-rc4 and 2.1.0-rc3
comparing 2.1.0-rc4 and 2.1.0-rc5
comparing 2.1.0-rc2 and 2.1.2
comparing 2.1.0-rc2 and 2.1.0-rc5
comparing 2.1.0-rc2 and 2.1.0-rc3
comparing 2.1.0-rc2 and 2.1.0-rc1
comparing 2.1.0-beta2 and 2.1.2
comparing 2.1.0-beta2 and 2.1.0-rc5
comparing 2.1.0-beta2 and 2.1.0-rc2
comparing 2.1.0-beta2 and 2.1.0-rc1
comparing 2.1.0-beta2 and 2.1.0-deb
comparing 2.1.0-beta1 and 2.1.1
comparing 2.1.0-beta1 and 2.1.0-rc4
comparing 2.1.0-beta1 and 2.1.0-rc1
comparing 2.1.0-beta1 and 2.1.0-deb
comparing 2.1.0-beta1 and 2.1.0-beta2
comparing 3.0.0-beta1 and 3.0.0-beta2
comparing 3.0.0-rc1 and 3.0.0-beta2
comparing 3.0.0-alpha1 and 3.0.0-beta2
comparing 3.0.0-alpha1 and 3.0.0-beta1
comparing 3.0.0-rc2 and 3.0.0-beta2
comparing 3.0.0-rc2 and 3.0.0-rc1
comparing 3.0.0 and 3.0.0-beta2
   special compare result: 1
comparing 3.0.0 and 3.0.0-rc2
   special compare result: 1

**Comment by knifewine** [comment]
@mambocab can you have another look?

**Comment by mambocab** [comment]
Code looks good to me, and the output looks correct. Could you run the same tests, but with the arguments reversed? If those work, then I'm +1.

**Comment by knifewine** [comment]
Did a bit more hacky testing and think we're good here. Thanks a bunch @mambocab !

**Pull request merged**